main/src/ca/uwaterloo/flix/api/lsp/provider/Visitors.scala

```diff
@@ -0,0 +1,316 @@
+package ca.uwaterloo.flix.api.lsp.provider
+
+import ca.uwaterloo.flix.api.lsp.Entity
+import ca.uwaterloo.flix.language.ast.TypedAst.Root
+import ca.uwaterloo.flix.language.ast.SourceLocation
+import ca.uwaterloo.flix.language.ast.TypedAst.Def
+import ca.uwaterloo.flix.language.ast.TypedAst.Expr
+import ca.uwaterloo.flix.language.ast.Type
+import ca.uwaterloo.flix.api.lsp.Position
+import ca.uwaterloo.flix.language.ast.TypedAst.Effect
+import ca.uwaterloo.flix.language.ast.Ast.Annotation
+import ca.uwaterloo.flix.language.ast.TypedAst.Constraint
+import ca.uwaterloo.flix.language.ast.TypedAst.Pattern
+import ca.uwaterloo.flix.language.ast.TypedAst.StructField
+
+
+object Visitors {
+  def visitRoot(visit: Entity => Unit, accept: SourceLocation => Boolean)(root: Root): Unit = {
+    // visit(Entity.Root(root))?
+
+    root.defs.map{ case (_, defn) => {
+      if (inside(accept, defn)) {
+        visitDef(visit, accept)(defn)
+      }
+    }}
+
+    root.effects.map{ case (_, v) => {
+      if (inside(accept, v)) {
+        visitEffect(visit, accept)(v)
+      }
+    }}
+
+    // root.entryPoint.map{ case v => visitEntryPoint(visit, accept)(v) }
+    // root.enums.map{ case (_, v) => visitEnum(???, ???)(v) };
+    // root.instances.map { case (_, v) => visitInstance(???, ???)(v) };
+    // root.modules.map { case (_, v) => visitModule(???, ???)(v) };
+    // root.restrictableEnums.map { case (_, v) => visitRestrictable(???, ???)(v) }; //experimental, prob should be removed
+    // root.sigs.map{ case (_, v) => visitSig(???, ???)(v) };
+    // root.structs.map{ case (_, v) => visitStruct(???, ???)(v) };
+    // root.traitEnv.map{ case (_, v) => visitTraitEnv(???, ???)(v) };
+    // root.traits.map{ case (_, v) =>  visitTrait(???, ???)(v) };
+    // root.typeAliases.map{ case (_, v) => visitTypeAlias(???, ???)(v) };
+    // root.uses.map{ case (_, v) => visitUse(???, ???)(v) };
+  }
+
+  def inside(uri: String, pos: Position)(loc: SourceLocation): Boolean = {
+    val posLine = pos.line + 1
+    val posCol = pos.character + 1
+
+    // sp1 and sp2, by invariant, has the same source, so we can use either
+    (uri == loc.sp1.source.name) &&
+    (posLine >= loc.beginLine) &&
+    (posLine <= loc.endLine) &&
+    !(posLine == loc.beginLine && posCol < loc.beginCol) &&
+    !(posLine == loc.endLine && posCol > loc.endCol)
+  }
+
+  def inside(other: SourceLocation => Boolean, defn: Def): Boolean = {
```

Comment:
Why is this function curried? And why does it take a Def?

Reply:
Just define _one function_ on locs and use that everywhere.

Reply:
ive removed all the higher order inside functions. Now it's just the uri-pos and loc ones that are left. Since we don't have loc at the top level and we kinda just let the partially applied `inside` using uri and pos propegate down, we actually never use the one that takes in two locs